### PR TITLE
Change for the condition when refund button is available

### LINKED ISSUE #6641: [Order Details] Only show Issue Refund for paid orders which have not been fully refunded
If the order is paid, but not fully refunded, show the Issue Refund button.

Note that currently the button is hidden once a refund happens, even if that's not a full refund.

This, combined with other tickets in https://github.com/orgs/woocommerce/projects/68/ should lead to most orders showing either the `Collect Payment` or the `Issue Refund` button.

However, partially paid orders will display both. ⚠⚠⚠ Under discussion. p91TBi-8P0-p2#comment-10270

- [x] Continue to show Issue Refund button for paid orders which have not been fully refunded
- [x] Ensure that the Issue Refund button shows for orders with only Shipping lines (left) to refund

### PR DESCRIPTION
**Pull request by kidinov**
Closes: #6641

### Description

The PR adds a condition that the refund button is visible only when an order is partially or fully paid

@anitaa1990 @hichamboushaba asking you to review this too, because this changes the logic which you were working with, so we might be missing something here

### Testing instructions

Try the following:
* Order is not paid -> NOT visible
* Order is fully paid -> Visible 
* Order is partially refunded -> Visible
* Order is fully refunded -> NOT Visible

~~Still in consultations with @joshheald about the state when the order is partially paid. I don't know how to end up in this state, therefore do not merge label~~

Apparently we don't support this at the moment in the app anyway, so we ignore this case for now

### Ensure that the Issue Refund button shows for orders with only Shipping lines (left) to refund
image

### CODE REVIEW

**Review by malinajirka — COMMENTED** [review]
Thanks for implementing this logic @kidinov. This new behavior finally makes sense to me :).

> > Still in consultations with @joshheald about the state when the order is partially paid. I don't know how to end up in this state, therefore do not merge label

> Apparently we don't support this at the moment in the app anyway, so we ignore this case for now

👍 If I recall correctly from my support rotation, partially payments are supported only with a separate extension.

-----------

One possible issue:
1. Open an order with "Pending payment" status (unpaid).
2. Change it's status to "Completed".
3. Notice the "Collect Payment" button disappears but the "Issue refund" button remains hidden.

------------

P.S. We might want to consider creating a test suite in a follow up PR? The logic is getting complicated over time and the relations between the fields could bite us if we don't test it properly. Wdyt?

**Comment by kidinov** [comment]
@malinajirka thank you for the review

> One possible issue:

Fixed it here, not sure if it's a good fix. Apparently optimistic update based on the assumption that status change does not have any side effects on the order model, but this is not true, e.g. `complete` makes order `paid` on the backend. @wzieba could you please verify if this is ok?
1a01ca8637c3bc21a82ad56f406edce86e4cea4b

> P.S. We might want to consider creating a test suite in a follow up PR? The logic is getting complicated over time and the relations between the fields could bite us if we don't test it properly. Wdyt?

Added some tests here 4fe577622d0b6d111c7c2e8d16a4b6913d93d78b

**Review by malinajirka — APPROVED** [review]
Thanks so much for updating the code and adding the tests. I'm really sorry the review took so long. Looks good to me.


I'll wait for @wzieba's double check just to make sure we are not missing any context.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt [hunk 1]
```diff
@@ -468,6 +468,7 @@ final class OrderDetailViewModel @Inject constructor(
                                         prepareTracksEventsDetails(result.event)
                                     )
                                 } else {
+                                    reloadOrderDetails()
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt]: As `reloadOrderDetails()` happens regardless of any conditions, I suggest moving this outside `when` to increase readability.

**Review by wzieba — APPROVED** [review]
> Apparently optimistic update based on the assumption that status change does not have any side effects on the order model, but this is not true, e.g. complete makes order paid on the backend.

Good point @kidinov - this scenario is an argument against optimistic updates in our app. As we can't determine the correct state from only local changes, we're prone to showing an invalid state :(.

Codewise it looks good - I have one comment.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt (hunk 1 above)
Comment (kidinov) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailViewModel.kt]: Good point! Done here bb260f516848708711309869f9ff337d35059939

**Pull request merged**